### ISSUE TITLE
Backport kernel commits to improve responsiveness

### ISSUE THREAD

**Issue by unclejack** [ValveSoftware/SteamOS#954]
#### Your system information

* Steam client version: not relevant
* SteamOS version: 3.4.2 (the version I've tested the patched kernel with)
* Opted into Steam client beta?: No
* Opted into SteamOS beta?: No 
* Have you checked for updates in Settings > System?: Yes

#### Please describe your issue in as much detail as possible:
These two commits make the system more responsive in scenarios in which the CPU usage is nowhere near 100% on all threads. They're the kind of small changes which should reduce stuttering slightly.

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.1&id=a8fb40966f19ff81520d9ccf8f7e2b95201368b8 applies as it is. This avoids redundant cache flushing. The change has been made to treat these AMD CPUs in a similar manner to the Intel CPUs.

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.1&id=aebef63cf7ff222660fd321c47e24452e8437673 makes changes to enable MWAIT for AMD CPUs which support MWAIT. The attached version of this patch makes the same change for the 5.13 kernel tree used by the Steam Deck.
[5.13-x86-Remove-vendor-checks-from-prefer_mwait_c1_over_halt.patch](https://github.com/ValveSoftware/SteamOS/files/10325924/5.13-x86-Remove-vendor-checks-from-prefer_mwait_c1_over_halt.patch)

Please test these changes. Valve probably has more hardware and more thorough QA procedures. The first commit has been brought into the upstream kernel tree a while ago.

I've tested the changes on my Steam Deck with a custom SteamOS 3.4.2. Games were running just fine with these changes. My external SSD doesn't seem to be able to boot the upstream SteamOS anymore, not even without changes. This prevents me from testing this further on the Steam Deck.

#### Steps for testing:

1. Build a kernel with the patches applied 
2. Boot SteamOS with the new kernel
3. Run games to make sure there are no regressions (stuttering, high CPU usage, crashes)
4. Watch videos, run games which use 5-20% of the CPU and other apps which aren't demanding

**Comment by gregkh** [comment]
Why not ask that the above commits be added to the upstream stable kernel releases to ensure that they automatically get applied in future updates?  Please read the kernel documentation at https://www.kernel.org/doc/html/latest/process/stable-kernel-rules.html for how to do this properly.

**Comment by unclejack** [comment]
@gregkh: You are right. These should be at least in the 5.15 LTS kernel. I haven't built a kernel based on the LTS 5.10 kernel tree yet. There are a few other commits which should be in those LTS kernel versions.

Since SteamOS is still using the 5.13 kernel, commits backported to 6.1.x, 5.15.x  and 5.10.x won't get backported implicitly to this custom kernel tree.

Perhaps SteamOS will move to the 6.1 LTS kernel in the near future.

**Comment by unclejack** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.1&id=a8fb40966f19ff81520d9ccf8f7e2b95201368b8 should certainly be backported to the 5.13 kernel which is used by the stable 3.x SteamOS.

The other patch might be something harder to accept for a stable kernel. It'll probably take another 6-12 months before we have a version of SteamOS with kernel 6.1 or newer.

Perhaps it's a good time to start looking into putting together an alternative to SteamOS for the Steam Deck which includes the kernel drivers, the fan control and power control.

**Comment by guilhermepiccoli** [comment]
Thanks a bunch for your suggestions @unclejack ! I just sent the first patch to [stable](https://lore.kernel.org/stable/20230125215145.1171151-1-gpiccoli@igalia.com/) (as per Greg's suggestion) and get it into the process of merging on 5.13.

Will analyze the other patch, that seems a bit more complex, but might end-up there as well.
Cheers!

**Comment by Plagman** [comment]
We're working on moving to 6.1 and are still working on getting that kernel to be fully caught up. Following that, our goal is to be able to just track upstream kernel releases for Deck without carrying many changes, if any.

**Comment by gregkh** [comment]
I've applied the submitted upstream commit to the stable kernels.

And if there's anything I can do to help out with moving to 6.1 or newer, just let me know!

**Comment by unclejack** [comment]
@gregkh @guilhermepiccoli : https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.8&id=e400ad8b7e6a1b9102123c6240289a811501f7d9 should probably be on the 5.15 LTS and older LTS kernels where it can be applied cleanly (or just with changes for formatting).

@guilhermepiccoli @Plagman Having the MGLRU [1] enabled for 6.1.x and newer would be very helpful when it comes to avoiding disk IO related stalls and should improve responsiveness further:
```
CONFIG_LRU_GEN=y
CONFIG_LRU_GEN_ENABLED=y
# CONFIG_LRU_GEN_STATS is not set
```

I've tested these changes with the linux-neptune-61-6.1.1.valve1-1 tree. SteamOS 3.4.4 works with this kernel. Unfortunately, the fan control doesn't seem to work. That issue isn't related to enabling the MGLRU. All games I've tested with it were stable on SteamOS 3.4.4 with the 6.1.1 kernel (both with the standard neptune config and with the MGLRU enabled; running from the microSD card; confirmed that the running kernel had MGLRU enabled).

I'd recommend enabling the MGLRU to test with it enabled. Arch Linux enables it for the default kernel. The software and the games I've tested so far have worked just fine on Arch Linux.

I'll keep testing the 6.1.x tree with SteamOS. Having the fan control fixed would help make that easier.

[1] https://lore.kernel.org/all/20220815071332.627393-15-yuzhao@google.com/ - MGLRU design overview

**Comment by guilhermepiccoli** [comment]
> @gregkh @guilhermepiccoli : https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.8&id=e400ad8b7e6a1b9102123c6240289a811501f7d9 should probably be on the 5.15 LTS and older LTS kernels where it can be applied cleanly (or just with changes for formatting).
> 

Hey @unclejack , thanks again for the ideas! This one is already on Deck's 5.13 (it was merged right after Linus merged in mainline), for some reason I was pretty sure it was sent to stable! Nut seems it [never happened](https://lore.kernel.org/stable/faa01372-07b0-3438-9305-123a3de9cc47@intel.com/) heh...

I'll sent it to 5.15/5.10, makes sense!

Regarding the Multi-gen LRU, very nice that you tested it - thanks for the feedback! I'll bring this discussion with the folks and hopefully we could enable it by default - should be there in fact, since it's on Arch already...

Cheers!

**Comment by unclejack** [comment]
@guilhermepiccoli: We'll want to backport https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v6.2-rc6&id=727209376f4998bc84db1d5d8af15afea846a92b to 6.1 as well if you've confirmed that it's OK. I've brought it up to make sure it's backported to 6.1, that the SteamOS kernel includes the patch and that SteamOS sets it up properly.

 > Regarding the Multi-gen LRU, very nice that you tested it - thanks for the feedback! I'll bring this discussion with the folks and hopefully we could enable it by default - should be there in fact, since it's on Arch already...
 
The generational LRU isn't enabled in the neptune config. You can check the config used by the 6.1.7 kernel package from the SteamOS repository. I had to enable it when building my own kernel.
 
I also need to use `CONFIG_X86_KERNEL_IBT=n` to build any 6.1.x kernel on Arch Linux right now due to https://lore.kernel.org/bpf/ff12051b-e8f9-44d7-a8ea-5fa7094375e6@app.fastmail.com/T/ 

> > @gregkh @guilhermepiccoli : https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.8&id=e400ad8b7e6a1b9102123c6240289a811501f7d9 should probably be on the 5.15 LTS and older LTS kernels where it can be applied cleanly (or just with changes for formatting).
> 
> Hey @unclejack , thanks again for the ideas! This one is already on Deck's 5.13 (it was merged right after Linus merged in mainline), for some reason I was pretty sure it was sent to stable! Nut seems it [never happened](https://lore.kernel.org/stable/faa01372-07b0-3438-9305-123a3de9cc47@intel.com/) heh...
> 
> I'll sent it to 5.15/5.10, makes sense!
> 
> Regarding the Multi-gen LRU, very nice that you tested it - thanks for the feedback! I'll bring this discussion with the folks and hopefully we could enable it by default - should be there in fact, since it's on Arch already...
> 
> Cheers!

**Comment by gregkh** [comment]
On Thu, Feb 02, 2023 at 10:43:41AM -0800, unclejack wrote:
> @guilhermepiccoli: We'll want to backport https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v6.2-rc6&id=727209376f4998bc84db1d5d8af15afea846a92b to 6.1 as well if you've confirmed that it's OK. I've brought it up to make sure it's backported to 6.1, that the SteamOS kernel includes the patch and that SteamOS sets it up properly.

This is already in the latest 6.1.y release, so no need to do anything.

thanks,

greg k-h

**Comment by guilhermepiccoli** [comment]
@unclejack as Greg said, it's already on 6.0/6.1, also it doesn't affect so far the AMD CPUs (like Deck's one), since the split lock detector is an Intel feature present only on modern Intel CPUs. Without the split lock detection stuff, we have no misery mode heh

About the Multi-gen LRU, I discussed internally and what happens is that the 6.1 kernel was built using a 6.0 config with "olddefconfig" mode, i.e., it had all 6.1 new stuff set to the default, and multi-gen LRU is default off!

The future 6.1 builds hopefully will use the proper Arch 6.1 config file, hence we'll pick that - thanks again for the suggestions.

**Comment by unclejack** [comment]
SteamOS 3.5 will have kernel 6.1. This one will most likely be maintained by Valve and the developers involved.

The commits which were recommended for backporting are included in the 6.1 kernel already. It's a good time to close this issue.

A new issue can be opened if some other helpful and not yet backported commits are found.
